### ISSUE TITLE
New listeners immediately after an update get stale results

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#1232]
Ran into this when upgrading from 3.3.1 to 4.3.1:

```dart
// replace with generated table
final someTables = {db.appSettings};

await db.customStatement('create table tbl (x int)');
await db.customInsert('insert into tbl values(1)');

Stream<int> watchValue() => db
  .customSelect('select * from tbl', readsFrom: someTables)
  .map((row) => row.read<int>('x'))
  .watchSingle();

expect(await watchValue().first, 1);
await Future.delayed(Duration.zero);

watchValue().listen(print); // (1)

// await Future.delayed(Duration.zero); // (2)
await db.customUpdate(
  'update tbl set x = 2',
  updates: someTables,
  updateKind: UpdateKind.update);
// await Future.delayed(Duration.zero); // (3)

expect(await watchValue().first, 2);
// ^ passes if you comment out 1, or unncomment 2, or uncomment 3
```

**Comment by simolus3** [comment]
The first `await watchValue().first` and awaiting the next event loop does not impact the test failing or not. Afterwards, what's happening is that

1. A listener attaches (via `watchValue().listen(print)`)
2. The stream starts running the select statement because the first listener has attached and we don't have any data
3. The update runs
4. Streams are invalidated 
5. The subscription for `expect(...)` attaches
6. The select statement from (2) completes. The stream adds that event now, which reaches both the print and the expectation

I actually just realized that moor uses a faulty mutex implementation, I'll get a hotfix out asap.

**Comment by simolus3** [comment]
Just published moor `4.3.2` which fixes this problem. Thanks for the report!

**Comment by Mike278** [comment]
Dang that's very subtle. Thanks so much for the record-speed fix!
